### PR TITLE
Exec backends drop 0

### PR DESCRIPTION
**Pull request by patrikfirek**

### CODE REVIEW

File: src/app/Exec/lib/backend_proxy.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    jenkins: JenkinsBackend
+};
+
+class BackendProxy extends BackendProxyBase {
+    constructor() {
+        super(Backend);
+    }
+
+    createExecutor(backend, data = null) {
+        const backendType = this.getBackend(backend).backend.backendType;
```
Comment (christensson) [inline src/app/Exec/lib/backend_proxy.js]: Add backendType getter to DirectBackend and JenkinsBackend...

File: src/app/Exec/lib/backend_proxy.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        if (Object.keys(this.executorClasses).indexOf(backendType) > -1) {
```
Comment (christensson) [inline src/app/Exec/lib/backend_proxy.js]: I think this is easier to read:
`this.executorClasses.hasOwnProperty(backendType)`

Or this:
`backendType in this.executorClasses`

File: src/app/Exec/lib/backend_proxy.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+            return new this.executorClasses[backendType](data);
+        }
+
+        throw Error(`Cannot find executor for backend '${backend}'`);
+    }
+
+    async start(config = {}, executorClasses, ...args) {
```
Comment (christensson) [inline src/app/Exec/lib/backend_proxy.js]: Why not have executorClasses part of config?
`this.executorClasses = Object.assign({}, config.executorClasses);`

File: src/app/UI/lib/managers/mgmt/client/components/sections/exec/Remove.js [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        console.log("RemoveLocal-RENDER", this.props);
+
+        return (
+            <TASection
+                breadcrumbs={this.props.breadcrumbs}
+                controls={this.props.controls}
+            >
+                <TAForm
+                    confirmText="Remove"
+                    primaryText="Remove code repository backend"
```
Comment (christensson) [inline src/app/UI/lib/managers/mgmt/client/components/sections/exec/Remove.js]: Update primaryText

File: src/app/Exec/lib/types/executor.js [hunk 5]
```diff
@@ -296,14 +152,11 @@ class Executor extends Type {
...
 
-    getEnv(jobInfo) {
-        return this._convToEnvObj(jobInfo, false, 6);
-    }
-
     async allocate(job, slave) {
         this.jobId = job._id;
         this.jobName = job.name;
         this.slaveId = slave._id;
+        this.backendId = slave.backend;
```
Comment (christensson) [inline src/app/Exec/lib/types/executor.js]: Shouldn't this be set at create? For example in BackendProxy#createExecutor where it is known.

File: src/app/Exec/lib/backend_proxy.js [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    direct: DirectBackend,
+    jenkins: JenkinsBackend
+};
+
+class BackendProxy extends BackendProxyBase {
+    constructor() {
+        super(Backend);
+    }
+
+    createExecutor(backend, data = null) {
```
Comment (christensson) [inline src/app/Exec/lib/backend_proxy.js]: I think that this method should set backendId/backend on created executor!
Set data to empty object by default. And further down below, do
```js
const createData = Object.assign({ backendId: backend }, data);
return new this.executorClasses[backendType](createData);
```

File: src/app/Exec/lib/backend_proxy.js (hunk 2 above)
Comment (patrikfirek) [inline src/app/Exec/lib/backend_proxy.js]: in operator doesn't work for arrays

File: src/app/Exec/lib/backend_proxy.js (hunk 1 above)
Comment (patrikfirek) [inline src/app/Exec/lib/backend_proxy.js]: done

File: src/app/Exec/lib/backend_proxy.js (hunk 2 above)
Comment (patrikfirek) [inline src/app/Exec/lib/backend_proxy.js]: done

File: src/app/Exec/lib/control.js [hunk 7]
```diff
@@ -350,7 +351,7 @@ class Control {
         }
 
         // Allocate executor
-        let executor = new Executor();
+        let executor = BackendProxy.instance.createExecutor(leastUtlizedSlave.backend);
```
Comment (patrikfirek) [inline src/app/Exec/lib/control.js]: not done; executor is set to null further down

File: src/app/Exec/lib/types/executor.js [hunk 8]
```diff
@@ -49,21 +31,14 @@ const getPathType = (pathStr) => {
...
-};
-
+// This class is a base abstract class
 class Executor extends Type {
     constructor(data) {
         super();
 
         this.jobId = false;
         this.slaveId = false;
+        this.backendId = false;
```
Comment (patrikfirek) [inline src/app/Exec/lib/types/executor.js]: done

File: src/app/Exec/lib/types/executor.js [hunk 9]
```diff
@@ -74,8 +49,6 @@ class Executor extends Type {
         this.finished = false;
         this.executing = false;
 
-        this.__ssh = new SshClient();
-
         if (data) {
             delete data.ssh;
             delete data.com;
```
Comment (patrikfirek) [inline src/app/Exec/lib/types/executor.js]: done

File: src/app/Exec/lib/backend_proxy.js (hunk 6 above)
Comment (patrikfirek) [inline src/app/Exec/lib/backend_proxy.js]: done

File: src/app/UI/lib/managers/mgmt/client/components/sections/exec/Remove.js (hunk 4 above)
Comment (patrikfirek) [inline src/app/UI/lib/managers/mgmt/client/components/sections/exec/Remove.js]: done

**Pull request merged**